va/validation-authority.go

```diff
@@ -169,18 +239,46 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 		// We don't expect to make multiple requests to a client, so close
 		// connection immediately.
 		DisableKeepAlives: true,
+		// Intercept Dial in order to connect to the IP address we
+		// select.
+		Dial: dialer.Dial,
 	}
+
 	logRedirect := func(req *http.Request, via []*http.Request) error {
-		va.log.Info(fmt.Sprintf("validateSimpleHTTP [%s] redirect from %q to %q", identifier, via[len(via)-1].URL.String(), req.URL.String()))
+		if len(challenge.ValidationRecord.SimpleHTTP) >= maxRedirect {
+			return fmt.Errorf("Too many redirects")
+		}
+
+		host := req.URL.Host
+		port = "80"
+		if va.TestMode {
+			port = "5001"
+		}
+		if strings.Contains(host, ":") {
+			splitHost := strings.SplitN(host, ":", 2)
+			if len(splitHost) <= 1 {
+				return fmt.Errorf("Malformed host")
+			}
+			host, port = splitHost[0], splitHost[1]
```

Comment:
I've kept the `:port` logic in the redirect function since that's the only place it is used but I've moved various other port logic into `resolveAndConstructDialer` since that should simplify some things elsewhere.